src/main/java/org/apache/mesos/scheduler/plan/DefaultBlock.java

```diff
@@ -0,0 +1,161 @@
+import org.apache.mesos.offer.OfferRequirement;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+
+/**
+ * This class is a default implementation of the Block interface.
+ */
+public class DefaultBlock implements Block {
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+    private final String name;
+    private final Optional<OfferRequirement> offerRequirementOptional;
+    private final UUID id = UUID.randomUUID();
+    private Status status;
+    private Map<Protos.TaskID, Status> tasks = new HashMap<>();
+
+    public DefaultBlock(String name, OfferRequirement offerRequirement, Status status) {
+        this(name, Optional.of(offerRequirement), status);
+    }
+
+    private DefaultBlock(String name, Optional<OfferRequirement> offerRequirementOptional, Status status) {
+        this.name = name;
+        this.offerRequirementOptional = offerRequirementOptional;
+        this.status = status;
+    }
+
+    private boolean isStatus(Status status) {
+        return status.equals(this.status);
+    }
+
+    private void setTaskIds(Collection <Protos.Offer.Operation> operations) {
+        tasks.clear();
+
+        for (Protos.Offer.Operation operation : operations) {
+            if (operation.getType().equals(Protos.Offer.Operation.Type.LAUNCH)) {
+                for (Protos.TaskInfo taskInfo : operation.getLaunch().getTaskInfosList()) {
+                    tasks.put(taskInfo.getTaskId(), Status.IN_PROGRESS);
+                }
+            }
+        }
+    }
+
+    private void setStatus(Status newStatus) {
+        Status oldStatus = status;
+        status = newStatus;
+        logger.info(getName() + ": changed status from: " + oldStatus + " to: " + newStatus);
+    }
+
+    @Override
+    public boolean isComplete() {
+        return isStatus(Status.COMPLETE);
+    }
+
+    @Override
+    public boolean isPending() {
+        return isStatus(Status.PENDING);
+    }
+
+    @Override
+    public boolean isInProgress() {
+        return isStatus(Status.IN_PROGRESS);
+    }
+
+    @Override
+    public Optional<OfferRequirement> start() {
+        return offerRequirementOptional;
+    }
+
+    @Override
+    public void updateOfferStatus(Optional<Collection<Protos.Offer.Operation>> operationsOptional) {
+        if (operationsOptional.isPresent()) {
+            setTaskIds(operationsOptional.get());
+            setStatus(Status.IN_PROGRESS);
+        }
+    }
+
+    @Override
+    public void restart() {
+        logger.warn("Restarting block: " + getName());
+        setStatus(Status.PENDING);
+    }
+
+    @Override
+    public void forceComplete() {
+        logger.warn("Forcing completion of block: " + getName());
+        setStatus(Status.COMPLETE);
+    }
+
+    @Override
+    public void update(Protos.TaskStatus status) {
+        if (!tasks.containsKey(status.getTaskId())) {
+            logger.warn(getName() + " ignoring irrelevant TaskStatus: " + status);
```

Comment:
I would expect this to occur on stale TaskStatus messages and so should be pretty rare, but it's a normal thing to occur so it's not worthy of an exception.  I can change to Info.  I put it as Warn because it's something worthy of investigation.  Definitely if you're seeing it a lot.  Let me know what you think.